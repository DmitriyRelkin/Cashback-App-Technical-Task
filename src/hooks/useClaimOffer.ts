import {useEffect} from "react";
import { fetchCashbackStart, fetchCashbackSuccess, fetchCashbackFailure } from '../store/slices/cashbackSlice';
import {useDispatch} from "react-redux";
import {Alert} from "react-native";

// Mock fetch function
const mockFetch = () => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      const isSucceed = Math.floor(Math.random() * 3) > 1;
      if (isSucceed) {
        Alert.alert('Success', 'Offer claimed successfully!');
        resolve({status: 200})
      } else {
        Alert.alert('Error', 'Failed to claim the offer!');
        reject()
      }

    }, 1000); // Simulate a network delay
  });
};

export const useFetchClaimOffer = (offerId) => {
  const dispatch = useDispatch();

  const fetchOffer = async () => {
    dispatch(fetchCashbackStart());
    try {
      const response = await mockFetch();
      if (response.status === 200) {
        dispatch(fetchCashbackSuccess(offerId));
      }
    } catch (error) {
      dispatch(fetchCashbackFailure());
    }
  };

  return {
    fetchOffer,
  }
}
