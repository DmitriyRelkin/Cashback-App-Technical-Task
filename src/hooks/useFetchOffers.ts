import {useEffect} from "react";
import { fetchOffersStart, fetchOffersSuccess, fetchOffersFailure } from '../store/slices/offersSlice';
import {useDispatch} from "react-redux";

// Mock fetch function
const mockFetch = () => {
  return new Promise((resolve) => {
    setTimeout(() => {
      const data = require('../../assets/data.json');
      resolve({
        json: () => Promise.resolve(data)
      });
    }, 1000); // Simulate a network delay
  });
};

export const useFetchOffers = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    const loadData = async () => {
      dispatch(fetchOffersStart());
      try {
        const response = await mockFetch();
        const data = await response.json();
        dispatch(fetchOffersSuccess(data.offers));
      } catch (error) {
        dispatch(fetchOffersFailure(error.toString()));
      }
    };

    loadData();
  }, [])
}
