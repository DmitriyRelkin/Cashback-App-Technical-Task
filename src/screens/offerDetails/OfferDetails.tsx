import React, {useState} from 'react';
import {Text, StyleSheet, Image, Alert, ScrollView, TouchableOpacity, View, ActivityIndicator} from 'react-native';
import {useFetchClaimOffer} from "../../hooks/useClaimOffer";
import {useSelector} from "react-redux";

const OfferDetails = ({ route }) => {
  const { offer } = route.params;
  const { retailerLogo, title, description, termsAndConditions, id } = offer;
  const { fetchOffer } = useFetchClaimOffer(id);

  const { data, loading, error } = useSelector((state: any) => state.cashback);

  const isClaimed = data.indexOf(id) !== -1;

  const claimOffer = () => {
    Alert.alert(
      'Confirm Claim',
      'Are you sure you want to claim this offer?',
      [
        {
          text: 'Cancel',
          style: 'cancel',
        },
        {
          text: 'OK',
          onPress: () => fetchOffer(),
        },
      ],
      { cancelable: true }
    );
  };

  if (loading) {
    return <ActivityIndicator style={styles.loader} size="large" />;
  }

  if (error) {
    return <Text style={styles.error}>Failed to load offers</Text>;
  }

  return (
    <ScrollView style={styles.container}>
      <Image source={{uri: retailerLogo}} style={styles.bannerImage} />
      <Text style={styles.title}>{title}</Text>
      <Text style={styles.description}>{description}</Text>
      <Text style={styles.terms}>{termsAndConditions}</Text>
      {!isClaimed ? (
        <TouchableOpacity style={styles.claimButton} onPress={claimOffer}>
          <Text style={styles.claimButtonText}>Claim Offer</Text>
        </TouchableOpacity>
      ) : (
        <View style={styles.appliedDescriptionContainer}>
          <Text style={styles.appliedDescription}>Cashback Applied!</Text>
        </View>
      )}
    </ScrollView>
  );
};

const styles = StyleSheet.create({
  container: {
    padding: 16,
  },
  bannerImage: {
    width: '100%',
    height: '100%',
    resizeMode: 'cover',
    borderRadius: 8,
  },
  title: {
    fontSize: 24,
    fontWeight: 'bold',
    marginVertical: 16,
  },
  sectionTitle: {
    fontSize: 20,
    fontWeight: 'bold',
    marginVertical: 8,
  },
  description: {
    fontSize: 16,
    marginVertical: 8,
  },
  terms: {
    fontSize: 16,
    color: '#555',
  },
  claimButton: {
    backgroundColor: '#007AFF',
    padding: 16,
    borderRadius: 8,
    alignItems: 'center',
    marginVertical: 16,
  },
  claimButtonText: {
    color: 'white',
    fontSize: 16,
    fontWeight: 'bold',
  },
  appliedDescriptionContainer: {
    marginVertical: 16,
    padding: 10,
    backgroundColor: 'green',
    borderRadius: 20,
  },
  appliedDescription: {
    textAlign: 'center',
    color: '#62ff00',
    fontSize: 16,
    fontWeight: 'bold',
  },
  loader: {
    flex: 1,
    justifyContent: 'center',
  },
  error: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    color: 'red',
  },
});

export default OfferDetails;
