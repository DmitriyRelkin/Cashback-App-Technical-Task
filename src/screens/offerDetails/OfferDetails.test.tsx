import React from 'react';
import renderer from 'react-test-renderer';
import { useSelector } from 'react-redux';
import { useFetchClaimOffer } from '../../hooks/useClaimOffer';
import OfferDetails from "./OfferDetails";

// Mock necessary hooks and components
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
}));

jest.mock('../../hooks/useClaimOffer', () => ({
  useFetchClaimOffer: jest.fn(),
}));

describe('OfferDetails', () => {
  const mockUseSelector = useSelector as jest.Mock;
  const mockUseFetchClaimOffer = useFetchClaimOffer as jest.Mock;

  const mockOffer = {
    retailerLogo: 'https://example.com/logo.png',
    title: 'Sample Offer',
    description: 'This is a sample offer description.',
    termsAndConditions: 'These are the terms and conditions.',
    id: 1,
  };

  const route = {
    params: {
      offer: mockOffer,
    },
  };

  const fetchOfferMock = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    mockUseFetchClaimOffer.mockReturnValue({
      fetchOffer: fetchOfferMock,
    });
  });

  it('renders the component correctly', () => {
    mockUseSelector.mockReturnValue({ data: [], loading: false, error: null });
    const tree = renderer.create(<OfferDetails route={route} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('displays a loader when loading', () => {
    mockUseSelector.mockReturnValue({ data: [], loading: true, error: null });
    const tree = renderer.create(<OfferDetails route={route} />);
    const loader = tree.root.findByType('ActivityIndicator');
    expect(loader).toBeTruthy();
  });

  it('displays an error message when there is an error', () => {
    mockUseSelector.mockReturnValue({ data: [], loading: false, error: 'Error' });
    const tree = renderer.create(<OfferDetails route={route} />);
    const errorMessage = tree.root.findByType('Text').children.includes('Failed to load offers');
    expect(errorMessage).toBeTruthy();
  });
});
