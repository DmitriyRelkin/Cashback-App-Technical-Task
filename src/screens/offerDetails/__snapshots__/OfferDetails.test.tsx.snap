// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfferDetails renders the component correctly 1`] = `
<RCTScrollView
  style={
    {
      "padding": 16,
    }
  }
>
  <View>
    <Image
      source={
        {
          "uri": "https://example.com/logo.png",
        }
      }
      style={
        {
          "borderRadius": 8,
          "height": "100%",
          "resizeMode": "cover",
          "width": "100%",
        }
      }
    />
    <Text
      style={
        {
          "fontSize": 24,
          "fontWeight": "bold",
          "marginVertical": 16,
        }
      }
    >
      Sample Offer
    </Text>
    <Text
      style={
        {
          "fontSize": 16,
          "marginVertical": 8,
        }
      }
    >
      This is a sample offer description.
    </Text>
    <Text
      style={
        {
          "color": "#555",
          "fontSize": 16,
        }
      }
    >
      These are the terms and conditions.
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#007AFF",
          "borderRadius": 8,
          "marginVertical": 16,
          "opacity": 1,
          "padding": 16,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Claim Offer
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
